### PR TITLE
In-memory engine: fix panic when write batch callback twice

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
fix panic when write batch callback twice
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by afeinberg — APPROVED** [review]
LGTM. /cc @tonyxuqqi to merge it.

File: components/hybrid_engine/src/write_batch.rs [hunk 1]
```diff
@@ -40,10 +45,13 @@ impl<EK: KvEngine> WriteBatch for HybridEngineWriteBatch<EK> {
     }
 
     fn write_callback_opt(&mut self, opts: &WriteOptions, mut cb: impl FnMut(u64)) -> Result<u64> {
+        let called = Arc::new(AtomicBool::new(false));
```
Comment (tonyxuqqi) [inline components/hybrid_engine/src/write_batch.rs:48]: Is Arc needed here?

**Review by tonyxuqqi — APPROVED** [review]

File: components/hybrid_engine/src/write_batch.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/hybrid_engine/src/write_batch.rs]: Good catch and done.

**Review by LykxSassinator — APPROVED** [review]

File: components/hybrid_engine/src/write_batch.rs [hunk 2]
```diff
@@ -40,10 +42,13 @@ impl<EK: KvEngine> WriteBatch for HybridEngineWriteBatch<EK> {
     }
 
     fn write_callback_opt(&mut self, opts: &WriteOptions, mut cb: impl FnMut(u64)) -> Result<u64> {
+        let called = AtomicBool::new(false);
         self.disk_write_batch
             .write_callback_opt(opts, |s| {
-                self.cache_write_batch.set_sequence_number(s).unwrap();
-                self.cache_write_batch.write_opt(opts).unwrap();
+                if !called.fetch_or(true, Ordering::SeqCst) {
```
Comment (overvenus) [inline components/hybrid_engine/src/write_batch.rs:48]: When does it write twice?

**Pull request merged**